zcash_client_backend/src/data_api.rs

```diff
@@ -225,14 +220,57 @@ pub trait WalletRead {
     /// or a UUID.
     type NoteRef: Copy + Debug + Eq + Ord;
 
+    /// Returns a list of spendable Sapling notes sufficient to cover the specified target value,
+    /// if possible.
+    fn select_spendable_sapling_notes(
+        &self,
+        account: AccountId,
+        target_value: Amount,
+        anchor_height: BlockHeight,
+        exclude: &[Self::NoteRef],
+    ) -> Result<Vec<ReceivedSaplingNote<Self::NoteRef>>, Self::Error>;
+}
+
+pub trait TransparentInputSource {
+    /// The type of errors produced by a wallet backend.
+    type Error;
+
+    /// Returns a list of unspent transparent UTXOs that appear in the chain at heights up to and
+    /// including `max_height`.
+    fn get_unspent_transparent_outputs(
+        &self,
+        address: &TransparentAddress,
+        max_height: BlockHeight,
+        exclude: &[OutPoint],
+    ) -> Result<Vec<WalletTransparentOutput>, Self::Error>;
+}
+
+pub trait InputSource:
+    SaplingInputSource<Error = <Self as InputSource>::Error>
+    + TransparentInputSource<Error = <Self as InputSource>::Error>
+{
+    type Error;
+}
+
+/// Read-only operations required for light wallet functions.
+///
+/// This trait defines the read-only portion of the storage interface atop which
+/// higher-level wallet operations are implemented. It serves to allow wallet functions to
+/// be abstracted away from any particular data storage substrate.
+pub trait WalletRead: InputSource<Error = <Self as WalletRead>::Error> {
+    type Error;
+
     /// Returns the height of the chain as known to the wallet as of the most recent call to
     /// [`WalletWrite::update_chain_tip`].
     ///
     /// This will return `Ok(None)` if the height of the current consensus chain tip is unknown.
-    fn chain_height(&self) -> Result<Option<BlockHeight>, Self::Error>;
+    fn chain_height(&self) -> Result<Option<BlockHeight>, <Self as WalletRead>::Error>;
```

Comment:
Is the `as WalletRead` qualification necessary? I would have thought this was unambiguous because of the `Error = <Self as WalletRead>::Error`. I.e. if all types that `Self::Error` could refer to are the same, it shouldn't be necessary to disambiguate (but I see why Rust might not actually have that rule).

Reply:
It appears that each trait's associated types are considered separately by the Rust compiler, and this is required. This would be understandable to me for the purpose of resolving conflicts when there's no inheritance relationship between the traits, but yeah I don't really understand why it requires this qualification here.

Reply:
I guess it's because there are multiple parent types and the type equalities that are enforced in the supertrait bounds are not sufficiently transparent to the compiler somehow.